### PR TITLE
Add countries to interaction journey - add validations

### PR DESCRIPTION
**Pull request by sekharpanja**
### Description of change
This is second of three PR set, adding export countries to interaction journey.
First one is: https://github.com/uktrade/data-hub-api/pull/2358

This PR handles all business rule validations:
* export countries can only be added to non-investment themed interactions. Hence `were_countries_discussed` is mandatory field for both `export` and `other` themed interactions.
* At least one country/status combination is mandatory if `were_countries_discussed` is set to True
* Same country is not allowed to be added more than once, be it with different status

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/core/validators/rules_based.py [hunk 1]
```diff
@@ -36,6 +37,26 @@ def field(self):
...
 
 
+class IsFeatureFlagActive(BaseRule):
+    """Rule to check if feature flag is active."""
+
+    def __call__(self, combiner) -> bool:
+        """
+        Returns True if the feature flag is active.
+        """
+        return is_feature_flag_active(self.field)
```
Comment (reupen) [inline datahub/core/validators/rules_based.py]: Using `field` to store the name of the feature flag doesn't seem ideal. Maybe we could override `__init__` and store it in a dedicated attribute?

File: datahub/interaction/serializers.py [hunk 2]
```diff
@@ -466,6 +472,35 @@ class Meta:
...
+                            'theme',
+                            [Interaction.THEMES.export, Interaction.THEMES.other],
+                        ),
+                    ),
+                ),
+                ValidationRule(
+                    'required',
+                    OperatorRule('export_countries', is_not_blank),
+                    when=AndRule(
+                        EqualsRule('were_countries_discussed', is_not_blank),
```
Comment (reupen) [inline datahub/interaction/serializers.py]: Is this rule needed? Seems like the next one would cover it (if it's True then it isn't blank).

File: datahub/interaction/serializers.py [hunk 3]
```diff
@@ -466,6 +472,35 @@ class Meta:
...
+                        EqualsRule('were_countries_discussed', True),
```
Comment (reupen) [inline datahub/interaction/serializers.py]: Small thing, but `OperatorRule('were_countries_discussed', bool)` would keep this consistent with the other rule.

File: datahub/interaction/test/views/test_interaction.py [hunk 4]
```diff
@@ -336,15 +347,157 @@ def test_add(self, extra_data, permissions):
...
+                    'service': Service.inbound_referral.value.id,
+                    'was_policy_feedback_provided': False,
+                    'communication_channel': partial(random_obj_for_model, CommunicationChannel),
+                    'were_countries_discussed': None,
+                },
+                {
+                    'were_countries_discussed': ['This field is required.'],
+                },
+            ),
+            (
```
Comment (reupen) [inline datahub/interaction/test/views/test_interaction.py]: I know the comment above is meant to cover this set of parameters, but could we add a comment above each one to make it clearer?

File: datahub/interaction/validators.py [hunk 5]
```diff
@@ -190,3 +190,35 @@ def __call__(self, data):
...
+
+        if not export_countries:
+            return
+
+        deduped = set()
+        for item in export_countries:
+            country = item['country']
+            if country in deduped:
+                raise serializers.ValidationError(
+                    "Same country can't be added more than once to export_countries.",
```
Comment (reupen) [inline datahub/interaction/validators.py]: Looks like the logic can be simplified e.g.:

```python
countries = [item['country'] for item in export_countries]
if len(countries) > len(set(countries)):
    raise serializers.ValidationError(...)
```

**Comment by reupen** [comment]
Some additional questions about the validation:

- if the feature flag is inactive, should we be making sure that the fields are blank to avoid letting unvalidated data through?
- do we need the validation that the export countries are unchanged during updates?

File: datahub/interaction/serializers.py (hunk 2 above)
Comment (sekharpanja) [reply datahub/interaction/serializers.py]: This would mean, we don't care if `export_countries` is blank when `were_countries_discussed` is either blank/null or False.

Comment (reupen) [reply datahub/interaction/serializers.py]: Yes – but isn't that covered by the next rule?

File: datahub/core/validators/rules_based.py [hunk 6]
```diff
@@ -36,6 +37,35 @@ def field(self):
         return self._field
 
 
+class IsFeatureFlagActive():
```
Comment (reupen) [inline datahub/core/validators/rules_based.py]: I think this should still inherit from `BaseRule` since `ValidationRule` etc. expect rules to be an instance of `AbstractRule`.

`field` is optional in `BaseRule.__init__()`, so I think it should be straightforward (i.e. we should be able to just add a call to `super().__init__()` if `BaseRule` is added to the list of base classes).

**Comment by reupen** [comment]
@sekharpanja, did you see these questions above?

> Some additional questions about the validation:
> 
>   * if the feature flag is inactive, should we be making sure that the fields are blank to avoid letting unvalidated data through?
> 
>   * do we need the validation that the export countries are unchanged during updates?

**Comment by sekharpanja** [comment]
> @sekharpanja, did you see these questions above?
> 
> > Some additional questions about the validation:
> > 
> > * if the feature flag is inactive, should we be making sure that the fields are blank to avoid letting unvalidated data through?
> > * do we need the validation that the export countries are unchanged during updates?

File: datahub/interaction/test/views/test_interaction.py [hunk 7]
```diff
@@ -9,10 +9,17 @@
 from rest_framework.reverse import reverse
 from rest_framework.settings import api_settings
 
-from datahub.company.test.factories import AdviserFactory, CompanyFactory, ContactFactory
-from datahub.core.constants import Service
+from datahub.company.models import Company, CompanyExportCountry
+from datahub.company.test.factories import (
```
Comment (elcct) [inline datahub/interaction/test/views/test_interaction.py]: Is this change intentional? The imports seem to fit in one line.

File: datahub/interaction/serializers.py [hunk 8]
```diff
@@ -100,6 +106,18 @@ class Meta:
         validators = []
 
 
+class InteractionExportCountrySerializer(serializers.ModelSerializer):
+    """
+    InteractionExportCountry serializer.
```
Comment (elcct) [inline datahub/interaction/serializers.py]: It looks like the doc string is short enough to be in one line.

File: datahub/interaction/serializers.py [hunk 9]
```diff
@@ -289,6 +319,32 @@ def _save_dit_participants(self, interaction, validated_dit_participants):
         for adviser in old_advisers - new_advisers:
             old_adviser_mapping[adviser].delete()
 
+    def _save_export_countries(self, interaction, validated_export_countries):
+        """
+        Adds export countries related to an interaction
+        Update is not allowed yet
```
Comment (elcct) [inline datahub/interaction/serializers.py]: Maybe it will be better to say that update is not implemented and an attempt to update will raise `NotImplementedError` exception. Also it would be great to add full stops, so we know the comment is whole.

File: datahub/interaction/test/views/test_interaction.py [hunk 10]
```diff
@@ -167,6 +211,118 @@ def test_add(self, extra_data, permissions):
...
+            'archived_by': None,
+            'archived_on': None,
+            'archived_reason': None,
+        }
+
+    @freeze_time('2017-04-18 13:25:30.986208')
+    @pytest.mark.parametrize('permissions', NON_RESTRICTED_ADD_PERMISSIONS)
+    def test_add_feature_flag_off(self, permissions):
+        """
+        Check export country fields are not allowed to be updated
```
Comment (elcct) [inline datahub/interaction/test/views/test_interaction.py]: Does this test do what comment says?

File: datahub/core/validators/rules_based.py [hunk 11]
```diff
@@ -266,6 +266,24 @@ def __call__(self, combiner) -> bool:
         return all(rule(combiner) for rule in self._rules)
 
 
+class NotRule(BaseRule):
+    """
+    Field-less NOT rule that can be test on a rule in the `when` argument in
+    `ValidationRule.__init__`.
+    """
```
Comment (reupen) [inline datahub/core/validators/rules_based.py]: Optional: I think it might be nice to make this pass through the field from `self._rule` rather than making this rule this field-less (so i.e. inherit from `AbstractRule` instead and define a property like:

```python
    @property
    def field(self):
        return self._rule.field
```
)

File: datahub/interaction/serializers.py [hunk 12]
```diff
@@ -500,6 +519,28 @@ class Meta:
...
+                        InRule(
+                            'theme',
+                            [Interaction.THEMES.export, Interaction.THEMES.other],
+                        ),
+                    ),
+                ),
+                ValidationRule(
+                    'invalid_for_update',
+                    OperatorRule('were_countries_discussed', is_blank),
+                    OperatorRule('export_countries', is_blank),
```
Comment (reupen) [inline datahub/interaction/serializers.py]: Don't we want to say here that the values shouldn't have changed, rather than the values are blank? (An added difficulty being that the order of the countries doesn't matter...)

File: datahub/interaction/serializers.py [hunk 13]
```diff
@@ -413,6 +482,65 @@ class Meta:
...
+                        ),
+                    ),
+                ),
+                ValidationRule(
+                    'invalid_when_no_countries_discussed',
+                    OperatorRule('export_countries', is_blank),
+                    when=AndRule(
+                        IsObjectBeingCreated(),
+                        IsFeatureFlagActive(INTERACTION_ADD_COUNTRIES),
+                        EqualsRule('were_countries_discussed', False),
```
Comment (reupen) [inline datahub/interaction/serializers.py]: Should we make this:

```suggestion
                        OperatorRule('were_countries_discussed', not_),
```
?

File: datahub/interaction/serializers.py [hunk 14]
```diff
@@ -413,6 +482,65 @@ class Meta:
...
+                    OperatorRule('were_countries_discussed', is_blank),
+                    OperatorRule('export_countries', is_blank),
+                    when=AndRule(
+                        IsObjectBeingCreated(),
+                        NotRule(IsFeatureFlagActive(INTERACTION_ADD_COUNTRIES)),
+                    ),
+                ),
+                ValidationRule(
+                    'invalid_for_investment',
+                    OperatorRule('were_countries_discussed', not_),
```
Comment (reupen) [inline datahub/interaction/serializers.py]: Do we need to add `OperatorRule('export_countries', not_)` here too?

File: datahub/interaction/serializers.py (hunk 12 above)
Comment (sekharpanja) [reply datahub/interaction/serializers.py]: Wouldn't it be better not to allow those fields during update?

Comment (reupen) [reply datahub/interaction/serializers.py]: It would probably be easier to return an error if they are specified in an update (PATCH request) at all, yep, but the rule will look at existing data too. So this is saying interactions with those fields set can only be edited if those fields are made blank.

It would probably be easier to define a `validate_were_countries_discussed` and `validate_export_countries` for this (the rules were only meant for cross-field validation).

File: datahub/interaction/test/views/test_interaction.py [hunk 15]
```diff
@@ -1137,38 +1737,197 @@ def test_draft_update_enforces_required_fields(self, permissions, data, error_re
...
+            (
+                {
+                    'were_countries_discussed': True,
                     'export_countries': [
                         {
                             'country': {
-                                'id': export_country.country.id,
-                                'name': export_country.country.name,
+                                'id': Country.greece.value.id,
+                                'name': Country.greece.value.name,
```
Comment (reupen) [inline datahub/interaction/test/views/test_interaction.py]: The name in `NestRelatedField`s isn't normally used in requests so would suggest leaving it out (to avoid confusion).

File: datahub/interaction/validators.py [hunk 16]
```diff
@@ -190,3 +190,32 @@ def __call__(self, data):
...
+    within list of export_countries.
+    """
+
+    def __init__(self):
+        """Initialises the validator."""
+        self.instance = None
+
+    def set_context(self, serializer):
+        """Saves a reference to the model object."""
+        self.instance = serializer.instance
```
Comment (reupen) [inline datahub/interaction/validators.py]: Was going to mention using DRF 3.11's `requires_context` when rebasing this, but actually it doesn't look like `self.instance` is used – can we remove these two methods?
